lxd/device/tpm.go

```diff
@@ -0,0 +1,248 @@
+package device
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/revert"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/subprocess"
+	"github.com/lxc/lxd/shared/validate"
+)
+
+type tpm struct {
+	deviceCommon
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *tpm) validateConfig(instConf instance.ConfigReader) error {
+	if !instanceSupported(instConf.Type(), instancetype.Container, instancetype.VM) {
+		return ErrUnsupportedDevType
+	}
+
+	rules := map[string]func(string) error{
+		"path": validate.IsNotEmpty,
+	}
+
+	err := d.config.Validate(rules)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to validate config")
+	}
+
+	return nil
+}
+
+// validateEnvironment checks if the TPM emulator is available.
+func (d *tpm) validateEnvironment() error {
+	// Validate the required binary.
+	_, err := exec.LookPath("swtpm")
+	if err != nil {
+		return fmt.Errorf("Required tool '%s' is missing", "swtpm")
+	}
+
+	if d.inst.Type() == instancetype.Container {
+		// Load module tpm_vtpm_proxy which creates the /dev/vtpmx device, required
+		// by the TPM emulator.
+		module := "tpm_vtpm_proxy"
+
+		err := util.LoadModule(module)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to load kernel module %q", module)
+		}
+	}
+
+	return nil
+}
+
+// Start is run when the device is added to the instance.
+func (d *tpm) Start() (*deviceConfig.RunConfig, error) {
+	err := d.validateEnvironment()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to validate environment")
+	}
+
+	tpmDevPath := filepath.Join(d.inst.Path(), fmt.Sprintf("tpm.%s", d.name))
+
+	if !shared.PathExists(tpmDevPath) {
+		err := os.Mkdir(tpmDevPath, 0755)
```

Comment:
Why 0755 and not 0700?

We should treat the TPM data as confidential so not seeing a good reason for it to be world readable.